test/jit/test_dtype_analysis.py

```diff
@@ -245,3 +320,87 @@ def func(input, weight, bias, y):
         y_val = torch.rand((1,), dtype=torch.float32)
         input_args = [conv_in.input, *conv_in.args, y_val]
         self.assert_dtype_equal_custom_args(func, input_args)
+
+
+class TestDtypeCustomRules(TestDtypeBase):
+    def assert_output_dtype_equal(self, expected_res, prop_graph):
+        actual_dtype = self.node_output_dtype(prop_graph)
+        if len(actual_dtype) == 1:
+            self.assert_tensor_dtype_equal(expected_res, actual_dtype[0])
+        else:
+            self.assertEqual(len(expected_res), len(actual_dtype))
+            for expected, actual in zip(expected_res, actual_dtype):
+                self.assert_tensor_dtype_equal(expected, actual)
+
+    def assert_tensor_dtype_equal(self, tensor_output, graph_dtype):
+        if not isinstance(tensor_output, torch.Tensor):
+            return
+        self.assertEqual(tensor_output.dtype, graph_dtype)
+
+    @ops([op for op in op_db if op.aten_name in custom_rules_works_list])
+    def test_custom_rules(self, device, dtype, op):
+        # Return fn, inputs_fn for all
+        sample_input = op.sample_inputs(device, dtype, requires_grad=False)[0]
+        input_args = [sample_input.input, *sample_input.args]
+        expected_res = op(*input_args, **sample_input.kwargs)
+        func = op.get_op()
+        traced_fn = create_traced_fn(self, func)
+
+        # Have to run the traced function to actually generate the trace
+        traced_fn(sample_input.input, *sample_input.args, **sample_input.kwargs)
+
+        # Run the Dtype Analysis
+        graph = traced_fn.graph  # Note this is a cached graph
+        input_tensors = [t for t in input_args if isinstance(t, torch.Tensor)]
+        self.prop_dtype_on_graph(graph, input_tensors)
+        self.assert_output_dtype_equal(expected_res, graph)
+
+    @ops([op for op in op_db if op.aten_name in custom_rules_works_list])
+    def test_custom_rules_ints(self, device, dtype, op):
+        # This is done because opinfos currently only runs on floats.
+        # Return fn, inputs_fn for all
+        sample_input = op.sample_inputs(device, dtype, requires_grad=False)[0]
+
+        if dtype == torch.float32:
+            dtype = torch.int32
+        else:
+            dtype = torch.int64
+
+        input_args = [sample_input.input, *sample_input.args]
+        expected_res = op(*input_args, **sample_input.kwargs)
+        func = op.get_op()
+        traced_fn = create_traced_fn(self, func)
+
+        # Have to run the traced function to actually generate the trace
+        traced_fn(sample_input.input, *sample_input.args, **sample_input.kwargs)
+
+        # Run the Dtype Analysis
+        graph = traced_fn.graph  # Note this is a cached graph
+        input_tensors = [t for t in input_args if isinstance(t, torch.Tensor)]
+        self.prop_dtype_on_graph(graph, input_tensors)
+        self.assert_output_dtype_equal(expected_res, graph)
+
+    @expectedFailure
```

Comment:
lmao, this is subtle, does it take a regex, so we could narrow down to our asserts?